project/scheduling/Interval.agda

```diff
@@ -33,73 +35,93 @@ record Interval : Set where
 ⌊_⌋ : Interval → ℕ
 ⌊_⌋ i = Interval.lower i
 
+module _ where
+  _≟Interval_ : Decidable {A = Interval} _≡_
+  [ lower ▹ size ] ≟Interval [ lower₁ ▹ size₁ ] with lower ≟ lower₁ | size ≟ size₁
+  ...                                              | yes p          | yes q = yes
+                                                                                (begin
+                                                                                 [ lower ▹ size ] ≡⟨ cong ([_▹_] lower) q ⟩
+                                                                                 [ lower ▹ size₁ ] ≡⟨ cong (λ z → [ z ▹ size₁ ]) p ⟩
+                                                                                 [ lower₁ ▹ size₁ ] ∎)
+  ...                                              | no ¬p          | yes q = no lem
+    where
+    lem : (x : [ lower ▹ size ] ≡ [ lower₁ ▹ size₁ ]) → ⊥
+    lem refl = ¬p refl
+  ...                                              | yes p          | no ¬q = no lem
+    where
+    lem : (x : [ lower ▹ size ] ≡ [ lower₁ ▹ size₁ ]) → ⊥
+    lem refl = ¬q refl
+  ...                                              | no ¬p          | no ¬q = no lem
+    where
+    lem : (x : [ lower ▹ size ] ≡ [ lower₁ ▹ size₁ ]) → ⊥
+    lem refl = ¬q refl
+
+instance
+  DecEqInterval : DecEq Interval
+  DecEq._≟_ DecEqInterval = _≟Interval_
+
 module _ where
   i₀ = [ 2 ▹ 4 ]
   i₁ = [ 4 ▹ 6 ]
   i₂ = [ 11 ▹ 5 ]
   i₃ = [ 1 ▹ 10 ]
 
-data _∈_ : (n : ℕ) → Interval → Set where
-  contains : {n : ℕ} → {i : Interval} → (proof : ((n ≤ ⌈ i ⌉) × (⌊ i ⌋ ≤ n))) → n ∈ i
+data _<<_ : Rel Interval 0ℓ where
+  strictly-less : {l r : Interval} → {proof : True (⌈ l ⌉ <? ⌊ r ⌋)} → l << r
```

Comment:
Came across this design pattern as a replacement for the checkSL design pattern that was being used before. Implementing these data types in terms of the True of their decidable proofs (instead of in terms of the proofs directly) allows agda to be able to compute their result without having to explicitly implement that checkSL-type function every time. For every relation or predicate that asserts some proof, I've implemented a function of the form `...⇒p` which explicitly shows the non-decidable version of the proof. I've found that this design also reads quite nicely at the type level. In the case of the "strictly less" data type, at the type level we have:

```agda
data _<<_ : Rel Interval 0ℓ

<<⇒p : {l r : Interval} → l << r → ⌈ l ⌉ < ⌊ r ⌋
```
and then we can directly write the following without needing to implement any other functions and agda is able to compute the proofs.

```agda
i₀ = [ 2 ▹ 4 ]
i₂ = [ 11 ▹ 5 ]

_ : i₀ << i₂
_ = strictly-less
``